src/subprocess-posix.cc

```diff
@@ -223,6 +337,12 @@ SubprocessSet::SubprocessSet() {
     Fatal("sigaction: %s", strerror(errno));
   if (sigaction(SIGHUP, &act, &old_hup_act_) < 0)
     Fatal("sigaction: %s", strerror(errno));
+
+  memset(&act, 0, sizeof(act));
+  act.sa_flags = SA_SIGINFO | SA_NOCLDSTOP;
+  act.sa_sigaction = SigChldHandler;
+  if (sigaction(SIGCHLD, &act, &old_chld_act_) < 0)
+    Fatal("sigaction: %s", strerror(errno));
```

Comment:
The signal should first be blocked before changing its signal handler. Add a `sigaddset(&set, SIGCHLD)` after line 327 above.

Note that the loop waiting in DoWork() uses `pselect() / ppoll()` specifically to unblock the signals atomically when waiting for events. This is why there is no reason to block it temporarily when spawning a new process.